### ISSUE TITLE
SNOW-739749: snowflake-sqlalchemy 1.4.4 requires sqlalchemy<2.0.0,>=1.4.0, but you have sqlalchemy 2.0.1 which is incompatible.

### ISSUE THREAD

**Issue by sfc-gh-jgriffith** [snowflakedb/snowflake-sqlalchemy#380]
Error when attempting to install snowflake-sqlalchemy with sqlalchemy 2.0. Error: 
`snowflake-sqlalchemy 1.4.4 requires sqlalchemy<2.0.0,>=1.4.0, but you have sqlalchemy 2.0.1 which is incompatible.`

Version 1.4.1 release notes say snowflake-sqlalchemy is now SQLAlchemy 2.0 compatible.

https://github.com/snowflakedb/snowflake-sqlalchemy/blob/7c8effdb8c38f37084d5165961df75cbbbc10ddf/setup.cfg#L50

Please answer these questions before submitting your issue. Thanks!

1. What version of Python are you using?

   Python 3.8.16
   

2. What operating system and processor architecture are you using?

   macOS-10.16-x86_64-i386-64bit

3. What are the component versions in the environment (`pip freeze`)?

   asn1crypto==1.5.1
certifi @ file:///private/var/folders/sy/f16zz6x50xz3113nwtb9bvq00000gp/T/abs_477u68wvzm/croot/certifi_1671487773341/work/certifi
cffi==1.15.1
charset-normalizer==2.1.1
cryptography==38.0.4
filelock==3.9.0
greenlet==2.0.2
idna==3.4
oscrypto==1.3.0
pycparser==2.21
pycryptodomex==3.17
PyJWT==2.6.0
pyOpenSSL==22.1.0
pytz==2022.7.1
requests==2.28.2
snowflake-connector-python==2.9.0
SQLAlchemy==2.0.1
typing_extensions==4.4.0
urllib3==1.26.14

4. What did you do?

   pip install snowflake-sqlalchemy

5. What did you expect to see?

   Successfully installed snowflake-sqlalchemy-1.4.4

6. Can you set logging to DEBUG and collect the logs?

**Comment by sfc-gh-aling** [comment]
hey @sfc-gh-jgriffith, we're aware of the V2 GA release of sqlalchemy.
We will plan time for testing it and do release accordingly.

**Comment by stuck1233333** [comment]
Do you have an estimated release timeframe for this?

**Comment by qwertystop** [comment]
Seconded the request for a timeframe, especially given that a previous version explicitly said it added compatibility.

**Comment by cladden** [comment]
We would like to use sqlalchemy 2.0 as well, but this is blocking - is there a release timeframe?

**Comment by cpcloud** [comment]
Any idea when sqlalchemy 2.0 support might land?

**Comment by cladden** [comment]
> hey @sfc-gh-jgriffith, we're aware of the V2 GA release of sqlalchemy.
> 
> We will plan time for testing it and do release accordingly.

@sfc-gh-aling do you have any updates on this?

**Comment by hellobrett** [comment]
Hi @sfc-gh-aling -- I see the first update to snowflake-sqlalchemy in 12+ months just went out!  Support for 2.0 yet!?  As of this morning, it's still restricted to <2.0 upon installation.

https://pypi.org/project/snowflake-sqlalchemy/

**Comment by benoneill-tenzo** [comment]
Would also quite like to know a release timeframe for 2.0 support :)

**Comment by cladden** [comment]
I was curious how close snowflake-sqlalchemy is from supporting sqlalchemy 2.0, so I manually installed it:

```
pip install --no-deps snowflake-sqlalchemy==1.4.6
```

It seems very close!

After some monkey-patching, I was able to create an engine and connect to snowflake:

```
def snowflake_sqlalchemy_20_monkey_patches():
    import sqlalchemy.util.compat

    # make strings always return unicode strings
    sqlalchemy.util.compat.string_types = (str,)
    sqlalchemy.types.String.RETURNS_UNICODE = True

    import snowflake.sqlalchemy.snowdialect

    snowflake.sqlalchemy.snowdialect.SnowflakeDialect.returns_unicode_strings = True

    # make has_table() support the `info_cache` kwarg
    import snowflake.sqlalchemy.snowdialect

    def has_table(self, connection, table_name, schema=None, info_cache=None):
        """
        Checks if the table exists
        """
        return self._has_object(connection, "TABLE", table_name, schema)

    snowflake.sqlalchemy.snowdialect.SnowflakeDialect.has_table = has_table

# usage: call this function before creating an engine:
snowflake_sqlalchemy_20_monkey_patches()
```

@sfc-gh-aling and snowflake team, for your 2.0 porting efforts, maybe 👆is helpful in assessing the gaps?

**Comment by rosomri** [comment]
@sfc-gh-aling Any updates?

**Comment by jamesroseman** [comment]
@sfc-gh-aling SQLAlchemy 2.0 support was supposedly added last August. Any reason why this still doesn't work 234 days later?

![Screenshot 2023-04-14 at 13 51 33](https://user-images.githubusercontent.com/1159190/232048552-5027a548-54d5-44a9-8755-b9e7f320b43b.png)

**Comment by hellobrett** [comment]
@jamesroseman From our Snowflake rep: The August release note was a typo.  It should have read, 

>snowflake-sqlalchemy ***syntax*** is now v2.0 compatible

**Comment by Econ808** [comment]
> @jamesroseman From our Snowflake rep: The August release note was a typo. It should have read,
> 
> > snowflake-sqlalchemy _**syntax**_ is now v2.0 compatible

Funny - at least I am not the only one who was sure that it implied that snowflake-sqlalchemy was ready for sqlalchemy>=2.0.0

Can you at least provide a timeframe for when you expect snowflake-sqlalchemy to be up-to-date with sqlalchemy?

**Comment by devholland** [comment]
Any further news on this? We'd love to start using sqlalchemy 2.0 with Snowflake, too.

**Comment by oliverlambson** [comment]
Bump?

**Comment by mikwieczorek** [comment]
Is there any update on supporting SQL-Alchemy >= 2.0?

**Comment by danb27** [comment]
Any update on this timeline?

**Comment by juandiegopalomino** [comment]
Latest langchain version also needs sqlalchemy 2.0 to function. It's getting in the way of our openAI adoption 😅 
<img width="872" alt="Screenshot 2024-02-14 at 4 20 56 PM" src="https://github.com/snowflakedb/snowflake-sqlalchemy/assets/10430635/a3c7ba52-2e2c-473e-af41-a917735d165d">

**Comment by danb27** [comment]
Bump. 

Can we get an update? This is becoming more and more of a problem.

**Comment by 3065923** [comment]
How do we still have no update???

**Comment by jacobsieradzki** [comment]
+1 - SQLAlchemy 2 is essential for my project and this is becoming a blocker!

**Comment by sfc-gh-dszmolka** [comment]
hi folks, first of all thank you everyone so much for bearing with us while the SQLAlchemy 2.0 compatibility is implemented! 

I can confirm the implementation is currently in progress and we plan to release it by end of Q1 (April 2024). Please note this is not a committed-to date, just a rough estimation which is subject to change.

Will keep this thread posted on the progress.

**Comment by YarShev** [comment]
Is there an ETA of this issue?

**Comment by sfc-gh-dszmolka** [comment]
thank you all for your continued patience - no fixed ETA yet, but we're very close to making this available to everyone !

Speaking about which - if you are
* already a Snowflake customer, with private preview legal documents signed
* **and** willing to give the new functionality an early try **and also willing to provide detailed feedback**
please do reach out to your Account Team to take this for a spin.

after this private preview phase closes, we'll release the SQLAlchemy 2.0 support for everyone after such a long wait. Appreciate everyone your patience and bearing with us while the last steps are completed, really should not take much longer.

**Comment by jose-cdevx** [comment]
here i am, waiting
(thanks for your work btw, you are amazing! congrats to the team)
![imagen](https://media1.tenor.com/m/JPZy6DEo3EcAAAAd/cat-waiting-cat-wait.gif)

**Comment by YarShev** [comment]
@sfc-gh-dszmolka, it's been about 3 weeks since your last message. I was wondering if you could give all us an update?

**Comment by sfc-gh-dszmolka** [comment]
hey folks according to the latest information I have from the product team, we intend to run the private preview (see above) for a while to get more feedback and hopefully iron out wrinkles. I don't have any timeline set in stone unfortunately but asked the team with the source of truth and will update if there's any news.

Until the feature is available in public preview, you're more than welcome to participate in the private preview as mentioned above, if that's an option for you. Thank you all for the understanding here.

**Comment by astronautas** [comment]
Up! Any updates?

**Comment by sfc-gh-dszmolka** [comment]
hi folks - regarding any more granular timelines above which have been already shared, i really have nothing to say at this point, I wish I had, but then you'd already know it from here.

Regarding the above concern, thank you for sharing it. Should be a concern for Python 3.8 though; as [numpy 2 now requires Python 3.9-3.12](https://numpy.org/devdocs/release/2.0.0-notes.html#highlights) and on Python 3.9 both libraries are in the required version (pandas 2.2.2, numpy 2.0.0)

**Comment by YarShev** [comment]
#469 is closed. Should we consider this issue resolved?

**Comment by sfc-gh-dszmolka** [comment]
Indeed! I'll close this issue once the version which is compatible with SQLAlchemy 2.0 is finally released.

**Comment by YarShev** [comment]
Should we expect that version latest by autumn 2024 as mentioned above?

**Comment by sfc-gh-dszmolka** [comment]
still don't have a definite and committed-to date from the Product team but I believe the plans to approach the release might have changed.
Just saw https://github.com/snowflakedb/snowflake-sqlalchemy/pull/514 so hopefully release v1.6.0 which is compatible with SQLAlchemy 2.0 is released _way_ sooner than the original worst-case estimation of autumn 2024.

**Comment by YarShev** [comment]
Look forward to that version. Thanks!

**Comment by sfc-gh-dszmolka** [comment]
After such a long wait, [release v1.6.0](https://github.com/snowflakedb/snowflake-sqlalchemy/releases/tag/v1.6.0) is out ! This is the one bringing compatibility with SQLAlchemy 2.0 and can be right now downloaded directly from here. 

For some reason it's not available on `pip` yet (still 1.5.3 is the latest) , we're working on this and will update this same thread.

Thank you everyone again for bearing with us for such a long time !

**Comment by Mrfence97** [comment]
Hooray 🎉! Thanks to all the team for working on this.

**Comment by tvoipio** [comment]
Excellent news, thanks to the team! Could you kindly comment here when 1.6.0 is available on PyPI and conda-forge? Our engineers might look at me funny if deployments involve downloading packages from some GitHub repository... (and I completely understand there is some kind of a delay, just trying to avoid having to poll both channels to see when we can update)

**Comment by edgarrmondragon** [comment]
> Excellent news, thanks to the team! Could you kindly comment here when 1.6.0 is available on PyPI and conda-forge? Our engineers might look at me funny if deployments involve downloading packages from some GitHub repository... (and I completely understand there is some kind of a delay, just trying to avoid having to poll both channels to see when we can update)

The publishing workflow was broken and they seem to have fixed it in #520, so I think it's just a matter of them creating a 1.6.1 release now.

**Comment by edgarrmondragon** [comment]
It's out now 🙂 

https://pypi.org/project/snowflake-sqlalchemy/1.6.1/

**Comment by sfc-gh-dszmolka** [comment]
indeed ! aforementioned temporal issue with the release is now fixed, and v1.6.1 version of the library is now out.

This marks a big milestone, and again i would like to thank you all for your patience here ! Marking this issue as closed :)
